concepts/eventing-code-guidelines.md

```diff
@@ -0,0 +1,1147 @@
+	port := uint32(443)
+	isExternal := true
+	apiRule.Spec.Service = &apigatewayv1alpha1.Service{
+		Name:       &svcName,
+		Port:       &port,
+		Host:       &host,
+		IsExternal: &isExternal,
+	}
+}
+
+apiRule = NewAPIRule(
+	subscriptionWithoutWebhookAuth, 
+	WithPath
+)
+WithServiceBefore(host, svcName, apiRule) // this applies a side effect
+```
+</details>
+
+<details>
+	<summary>Do</summary>
+
+In the following example, `WithServiceAfter` is used as an argument to `NewAPIRule`.
+
+```go
+func WithServiceAfter(name, host string) APIRuleOption {
+	return func(r *apigatewayv1alpha1.APIRule) {
+		port := uint32(443)
+		isExternal := true
+		r.Spec.Service = &apigatewayv1alpha1.Service{
+			Name:       &name,
+			Port:       &port,
+			Host:       &host,
+			IsExternal: &isExternal,
+		}
+	}
+}
+
+apiRule = NewAPIRule(
+	subscriptionWithoutWebhookAuth,
+	WithPath(),
+	WithServiceAfter(svcName, host), // instead pass it as an argument to NewAPIRule
+)
+```
+</details>
+
+**Example 3**: Side effect as new function
+
+<details>
+  <summary>Don't</summary>
+
+`WithStatusReady` is not used inside a function that accepts a functional option as input.
+As a reader, you would not expect that it performs a **side-effect** on the `apiRuleNew`.
+
+```go
+// source: https://github.com/kyma-project/kyma/blob/a84f76f674babfc63e369bbba44de76499cf475d/components/eventing-controller/controllers/subscription/beb/reconciler_test.go#L592
+getAPIRule(ctx, apiRuleNew).Should(And(
+	HaveNotEmptyHost(),
+	HaveNotEmptyAPIRule(),
+))
+WithStatusReady(apiRuleNew) // <= WithStatusReady not used inside a function that accepts a functional option as input (for example, `APIRuleOption`)
+```
+
+</details>
+
+<details>
+  <summary>Do</summary>
+
+To make the side-effect more obvious, the following example uses a function named `MarkReady` instead.
+
+```go
+getAPIRule(ctx, apiRuleNew).Should(And(
+	HaveNotEmptyHost(),
+	HaveNotEmptyAPIRule(),
+))
+MarkReady(apiRuleNew) // instead consider using another function that explicitly states from the name that it applies a side effect
+```
+</details>
+
+## Testing pyramid in Kyma
+
+The following section describes the different testing levels for Kyma. These levels are usually seen as a [pyramid](https://en.wikipedia.org/wiki/Test_automation#Testing_at_different_levels):
+- At the bottom of the pyramid, you should have the **most** test cases because they are usually very **fast**. However, you are limited in what you can test because external dependencies are **mocked**. Unit tests are an example of this.
+- At the top of the pyramid, there are usually **fewer** tests because these tests take **longer**. However, these tests are a more realistic scenario of how the end user will use our product. E2E tests are an example of this.
+
+### Controller Integration Tests (kubebuilder)
+
+**Test Setup**
+Controllers using the kubebuilder framework bootstrap integration tests using `Ginkgo` (for testing) and `Gomega` (for assertions).
```

Comment:
that's a lot of "using" in one sentence. Can we rephrase/split the sentence so it's easier to understand?